server/lib/src/idm/authsession.rs

```diff
@@ -757,34 +757,41 @@ impl AuthSession {
             // and interact with the account more?
             if account.is_anonymous() {
-                AuthSessionState::Init(vec![CredHandler::Anonymous {
+                AuthSessionState::Init(nonempty![CredHandler::Anonymous {
                     cred_id: account.uuid,
                 }])
             } else {
                 // What's valid to use in this context?
-                let mut handlers = Vec::new();
-
-                if let Some(cred) = &account.primary {
-                    // TODO: Make it possible to have multiple creds.
-                    // Probably means new authsession has to be failable
-                    if let Ok(ch) = CredHandler::try_from((cred, webauthn)) {
-                        handlers.push(ch);
-                    } else {
-                        security_critical!(
-                            "corrupt credentials, unable to start primary credhandler"
-                        );
+                // The crab god forsake this place
+                match (
+                    &account.primary,
+                    CredHandler::try_from((&account.passkeys, webauthn)),
+                ) {
+                    (Some(cred), Ok(ch)) => {
+                        if let Ok(ch_cred) = CredHandler::try_from((cred, webauthn)) {
+                            AuthSessionState::Init(nonempty![ch_cred, ch])
+                        } else {
+                            security_critical!(
+                                "corrupt credentials, unable to start primary credhandler"
+                            );
+                            AuthSessionState::Init(nonempty![ch])
+                        }
+                    }
+                    (None, Ok(ch)) => AuthSessionState::Init(nonempty!(ch)),
+                    (Some(cred), Err(_)) => {
+                        if let Ok(ch_cred) = CredHandler::try_from((cred, webauthn)) {
+                            AuthSessionState::Init(nonempty![ch_cred])
+                        } else {
+                            security_critical!(
+                                "corrupt credentials, unable to start primary credhandler"
+                            );
+                            AuthSessionState::Denied("invalid credential state")
+                        }
+                    }
+                    _ => {
+                        security_info!("account has no available credentials");
+                        AuthSessionState::Denied("invalid credential state")
                     }
-                }
-
-                if let Ok(ch) = CredHandler::try_from((&account.passkeys, webauthn)) {
-                    handlers.push(ch);
-                };
-
-                if handlers.is_empty() {
-                    security_info!("account has no available credentials");
-                    AuthSessionState::Denied("invalid credential state")
-                } else {
-                    AuthSessionState::Init(handlers)
```

Comment:
I think a better idea is to use the traditional vec approach to build handlers, then to convert it to non-empty in this step. It keeps the code cleaner *and* still ends up with the non empty at the end.

Reply:
That approach might have its own ugliness since a non empty can't be straight collected from an iterator. I will try it out and see what it looks like.

Reply:
Actually we can use slice pattern here.

Reply:
Yeah, I think there are ways to do this nicely.